public/app/features/templating/template_srv.ts

```diff
@@ -239,56 +246,93 @@ export class TemplateSrv implements BaseTemplateSrv {
     if (!target) {
       return target ?? '';
     }
+    return this.targetTokens(target, scopedVars, format)
+      .map(t => this.formatValue(t.value, t.format, t.variable, t.text))
+      .join('');
+  }
 
-    this.regex.lastIndex = 0;
+  /**
+   * Returns all possible interpolated values for the given target.
+   * Each variable value is formatted individually,
+   * so formats are not used to join multiple values but
+   * only for single-value formatting, e.g. for escaping or text representation.
+   */
+  replaceToList(target: string, scopedVars?: ScopedVars, format?: string | Function): string[] {
+    return TemplateSrv.cartesianProduct(
+      this.targetTokens(target, scopedVars, format).map(t =>
+        (_.isArray(t.value) ? t.value : [t.value]).map(v => this.formatValue(v, t.format, t.variable, t.text))
+      )
+    ).map(t => t.join(''));
+  }
 
-    return target.replace(this.regex, (match, var1, var2, fmt2, var3, fieldPath, fmt3) => {
-      const variableName = var1 || var2 || var3;
-      const variable = this.getVariableAtIndex(variableName);
-      const fmt = fmt2 || fmt3 || format;
+  private static cartesianProduct(arr: any[][]) {
+    return arr.reduce((acc, curr) => acc.flatMap(a => curr.map(c => a.concat([c]))), [[]]);
```

Comment:
This is a mouthful indeed :)

Reply:
Yes it is. But it's one of the most readable solutions of https://stackoverflow.com/questions/12303989/cartesian-product-of-multiple-arrays-in-javascript
I added a comment for the function.